### ISSUE TITLE
Connection error 

### ISSUE THREAD

**Issue by Maryam-Imani** [openeventdata/mordecai#9]
I confronted with a connection error in the updated Mordecai. I did not have this issue with the previous Mordecai. I will be thankful if you can help me to solve this error. 

```
mordecai_1 | WARNING:elasticsearch:GET http://localhost:9200/_search [status:N/A request:0.000s]
mordecai_1 | Traceback (most recent call last):
mordecai_1 |   File "/usr/local/lib/python2.7/dist-packages/elasticsearch/connection/http_urllib3.py", line 74, in perform_request
mordecai_1 |     response = self.pool.urlopen(method, url, body, retries=False, headers=self.headers, **kw)
mordecai_1 |   File "/usr/local/lib/python2.7/dist-packages/urllib3/connectionpool.py", line 597, in urlopen
mordecai_1 |     _stacktrace=sys.exc_info()[2])
mordecai_1 |   File "/usr/local/lib/python2.7/dist-packages/urllib3/util/retry.py", line 222, in increment
mordecai_1 |     raise six.reraise(type(error), error, _stacktrace)
mordecai_1 |   File "/usr/local/lib/python2.7/dist-packages/urllib3/connectionpool.py", line 544, in urlopen
mordecai_1 |     body=body, headers=headers)
mordecai_1 |   File "/usr/local/lib/python2.7/dist-packages/urllib3/connectionpool.py", line 349, in _make_request
mordecai_1 |     conn.request(method, url, **httplib_request_kw)
mordecai_1 |   File "/usr/lib/python2.7/httplib.py", line 979, in request
mordecai_1 |     self._send_request(method, url, body, headers)
mordecai_1 |   File "/usr/lib/python2.7/httplib.py", line 1013, in _send_request
mordecai_1 |     self.endheaders(body)
mordecai_1 |   File "/usr/lib/python2.7/httplib.py", line 975, in endheaders
mordecai_1 |     self._send_output(message_body)
mordecai_1 |   File "/usr/lib/python2.7/httplib.py", line 835, in _send_output
mordecai_1 |     self.send(msg)
mordecai_1 |   File "/usr/lib/python2.7/httplib.py", line 797, in send
mordecai_1 |     self.connect()
mordecai_1 |   File "/usr/local/lib/python2.7/dist-packages/urllib3/connection.py", line 155, in connect
mordecai_1 |     conn = self._new_conn()
mordecai_1 |   File "/usr/local/lib/python2.7/dist-packages/urllib3/connection.py", line 134, in _new_conn
mordecai_1 |     (self.host, self.port), self.timeout, **extra_kw)
mordecai_1 |   File "/usr/local/lib/python2.7/dist-packages/urllib3/util/connection.py", line 88, in create_connection
mordecai_1 |     raise err
mordecai_1 | ProtocolError: ('Connection aborted.', error(111, 'Connection refused'))
```

**Comment by ahalterman** [comment]
Uh oh. Can you give me a reproducible example so I can try to figure out what's going on? I've run into a couple of similar errors that resulted from calling Mordecai before it was fully loaded. Does the problem persist after you try a couple of times?

**Comment by imani-ma** [comment]
I have this problem for all my dataset. However, I have attached one of them. 
[test.txt](https://github.com/openeventdata/mordecai/files/364654/test.txt)

The previous version gives the following result: 
[Output.txt](https://github.com/openeventdata/mordecai/files/364659/Output.txt). 

But the updated Mordecai has a connection error. So it does not have any output.

**Comment by ahalterman** [comment]
This looks like a problem with the Elasticsearch container. What's the output of `sudo docker ps`? And what happens when you do `curl -XGET http://localhost:9200/_search`?

**Comment by imani-ma** [comment]
$ sudo docker ps
CONTAINER ID        IMAGE                       COMMAND                  CREATED             STATUS              PORTS                    NAMES
b9753dc6cd32        mordecaimaster3_mordecai    "python /src/app.py"     47 hours ago        Up 31 minutes       0.0.0.0:5000->5000/tcp   mordecaimaster3_mordecai_1
188a73e53717        openeventdata/es-geonames   "/bin/sh -c 'cd elast"   47 hours ago        Up 31 minutes       9200/tcp, 9300/tcp       mordecaimaster3_elastic_1


 $ curl -XGET http://localhost:9200/_search
{"took":1,"timed_out":false,"_shards":{"total":0,"successful":0,"failed":0},"hits":{"total":0,"max_score":0.0,"hits":[]}}

**Comment by ahalterman** [comment]
I've made some progress on tracking down the error. It is indeed a problem with how the containers are linked using `docker-compose`. I'm thinking of moving away from `docker-compose` to straight `docker`. That would mean that users would have to download and run the `es-geonames` container, and then start `mordecai` with a link to the `es-geonames` container with something like this:
```
sudo docker run --link=es-geonames:elastic  -t -i 047b579a9aa8 /bin/bash
```
where `elastic` maps to the `es-geonames` container inside Mordecai and `047b579a9aa8` is the ID of the `mordecai` container.

`docker-compose` is very nice for setup, but not so great for ease of maintenance. Specific advantages:

1. a standalone geonames container is needed for running unit tests
2. the container could be useful for things outside of Mordecai
3. I think it's conceptually easier to have two separately started containers than linked containers, but maybe that's my own bias here.

**Comment by ahalterman** [comment]
After talking to @chilland, a Docker pro, it sounds like the issue is that `docker-compose` has changed significantly, so the original `docker-compose.yml` won't work. Unless anyone has strong objections, I'm going to de-compose Mordecai for the reasons above rather than revamp everything to use the new `docker-compose`.

**Comment by johnb30** [comment]
For a `docker-compose` 2 file it should look like this:

```
version: '2'
services:
    mordecai:
        build: .
        ports:
            - "5000:5000"
    elastic:
        image: openeventdata/es-geonames
```

You can look at [https://docs.docker.com/compose/networking/](https://docs.docker.com/compose/networking/) for info on how the networking works in the newer compose versions. The basic idea is that instead of `localhost` or `APPLICATION_PORT_5000_TCP_ADDR` you'd just used `elastic` as the host since that's the name that you define it as in the compose file. 

The way mordecai is setup is a bit weird right now since we were trying to account for the fact that someone might not want to run an ES server on the localhost. So, and @ahalterman I know this a bit different than what I've said in other conversations, it might make sense to split the ES server off from the `docker-compose` setup. It's rarely a good idea to run the ES server on localhost, I think, since it chews up a ton of RAM. Basically I'd be 50-50 either way. I think there are distinct benefits to leaving the ES server in, but can see an argument for pulling it out. I do strongly believe that splitting it into different pieces will make the user experience more difficult, though.

**Comment by PTB-OEDA** [comment]
Depends also in part with what arch / development environments we need to
support. There is the fixed thing like Cline and the streaming we want for
real time. How do we make this work for both? Or do we need separate
solutions?

**Comment by johnb30** [comment]
Streaming vs. batch applications don't really matter here. The scale would, instead, need to consider permanence of deployment. There is, as with currently all of the OEDA software, a difference between deploying something locally to test a few things and expecting to run X million observations through it.

**Comment by PTB-OEDA** [comment]
Now that I think about it, agreed. Let me know if we inputs from others
that I can coordinate.

**Comment by chilland** [comment]
Ideally all of these pieces would be composable and generally agnostic of application. I would include whatever backend gazeteer lookup mechanism you choose to use in that statement. You could still certainly provide docker-compose files for common use cases, but it is impossible to support all of them. Best to just make it as flexible as possible.

**Comment by chilland** [comment]
Just to add to that. You can support both deployment methods in the same way. If the database host is an arg, it can be whatever you want, including an address using docker networking. The overlay networking could still be used for multi-host or single host. Or you could deploy with out using docker at all.

**Comment by ahalterman** [comment]
@Maryam-Imani: Will you check if PR #12 resolves your issue? If so, and if no one else in this thread has an objection, I'll merge it to master tomorrow.

**Comment by chrisbennight** [comment]
I had the same issue as the described here, and tried the adm1 branch - build with the instructions in the readme;   

everything loads and seems to be networkable now, but the python api only returns an empty array for the exemplar queries -- example below:

(not sure if this should be a new issue or included in this, since it's against the adm1 branch)

OS is ubuntu 16.04

```
root@linux-dev:~# docker ps
CONTAINER ID        IMAGE                       COMMAND                  CREATED             STATUS              PORTS                              NAMES
ab51321158a6        mordecai                    "python /src/app.py"     7 minutes ago       Up 7 minutes        0.0.0.0:5000->5000/tcp             grave_lovelace
09ddb8ae1d2e        openeventdata/es-geonames   "/bin/sh -c 'cd elast"   22 hours ago        Up 22 hours         0.0.0.0:9200->9200/tcp, 9300/tcp   elastic
root@linux-dev:~#
root@linux-dev:~# curl -XPOST -H "Content-Type: application/json"  --data '{"text":"(Reuters) - The Iraqi government claimed victory over Islamic State insurgents in Tikrit on Wednesday after a month-long battle for the city supported by Shiite militiamen and U.S.-led air strikes, saying that only small pockets of resistance remained. State television showed Prime Minister Haidar al-Abadi, accompanied by leaders of the army and police, the provincial governor and Shiite paramilitary leaders, parading through Tikrit and raising an Iraqi flag. The militants captured the city, about 140 km (90 miles) north of Baghdad, last June as they swept through most of Iraqs Sunni Muslim territories, swatting aside a demoralized and disorganized army that has now required an uneasy combination of Iranian and American support to get back on its feet."}' 'http://localhost:5000/places'
[]
root@linux-dev:~# 
root@linux-dev:~# curl -XGET http://localhost:9200/_search
{"took":30,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":9455860,"max_score":1.0,"hits":[{"_index":"geonames","_type":"geoname","_id":"AU4wG06OSPt6k9TaGZ1e","_score":1.0,"_source":{"elevation": "", "name": "Tolsally\u014fng-t\u2019\u014fn\u014fl", "modification_date": "2014-01-01", "geonameid": "7584353", "feature_class": "R", "admin3_code": "", "admin2_code": "", "coordinates": "38.25169,128.09909", "cc2": "", "country_code3": "KOR", "country_code2": "KR", "feature_code": "TNL", "timzeone": "Asia/Seoul", "dem": "928", "admin1_code": "06", "asciiname": "Tolsallyong-t'onol", "alternativenames": "Tolsallyong-t'onol,Tolsally\u014fng-t\u2019\u014fn\u014fl,dolsanlyeongteoneol,\ub3cc\uc0b0\ub839\ud130\ub110", "admin4_code": "", "population": "0"}},{"_index":"geonames","_type":"geoname","_id":"AU4wG06OSPt6k9TaGZ1j","_score":1.0,"_source":{"elevation": "", "name": "S\u014frik-kol", "modification_date": "2014-01-01", "geonameid": "7584358", "feature_class": "T", "admin3_code": "", "admin2_code": "", "coordinates": "38.07845,128.1616", "cc2": "", "country_code3": "KOR", "country_code2": "KR", "feature_code": "VAL", "timzeone": "Asia/Seoul", "dem": "359", "admin1_code": "06", "asciiname": "Sorik-kol", "alternativenames": "Sorik-kol,S\u014frik-kol,solisgol,\uc18c\ub9bf\uace8", "admin4_code": "", "population": "0"}},{"_index":"geonames","_type":"geoname","_id":"AU4wG06OSPt6k9TaGZ1o","_score":1.0,"_source":{"elevation": "", "name": "Hapkang il-gyo", "modification_date": "2014-01-01", "geonameid": "7584363", "feature_class": "S", "admin3_code": "", "admin2_code": "", "coordinates": "38.09349,128.17968", "cc2": "", "country_code3": "KOR", "country_code2": "KR", "feature_code": "BDG", "timzeone": "Asia/Seoul", "dem": "200", "admin1_code": "06", "asciiname": "Hapkang il-gyo", "alternativenames": "Hapkang il-gyo,habgang1gyo,\ud569\uac151\uad50", "admin4_code": "", "population": "0"}},{"_index":"geonames","_type":"geoname","_id":"AU4wG06OSPt6k9TaGZ1t","_score":1.0,"_source":{"elevation": "", "name": "Nambukgyo", "modification_date": "2014-01-01", "geonameid": "7584368", "feature_class": "S", "admin3_code": "", "admin2_code": "", "coordinates": "38.06077,128.17031", "cc2": "", "country_code3": "KOR", "country_code2": "KR", "feature_code": "BDG", "timzeone": "Asia/Seoul", "dem": "197", "admin1_code": "06", "asciiname": "Nambukgyo", "alternativenames": "Nambuk-kyo,Nambukgyo,nambuggyo,\ub0a8\ubd81\uad50", "admin4_code": "", "population": "0"}},{"_index":"geonames","_type":"geoname","_id":"AU4wG06OSPt6k9TaGZ1y","_score":1.0,"_source":{"elevation": "494", "name": "Amisan", "modification_date": "2014-01-01", "geonameid": "7584373", "feature_class": "T", "admin3_code": "", "admin2_code": "", "coordinates": "38.04155,128.16385", "cc2": "", "country_code3": "KOR", "country_code2": "KR", "feature_code": "MT", "timzeone": "Asia/Seoul", "dem": "235", "admin1_code": "06", "asciiname": "Amisan", "alternativenames": "Amisan,Kunch'ung-nyong,Kunch\u2019ung-ny\u014fng,amisan,gunchuglyeong,\uad70\ucd95\ub839,\uc544\ubbf8\uc0b0", "admin4_code": "", "population": "0"}},{"_index":"geonames","_type":"geoname","_id":"AU4wG06OSPt6k9TaGZ13","_score":1.0,"_source":{"elevation": "", "name": "Kaeun-dong", "modification_date": "2014-01-01", "geonameid": "7584378", "feature_class": "T", "admin3_code": "", "admin2_code": "", "coordinates": "38.05107,128.12971", "cc2": "", "country_code3": "KOR", "country_code2": "KR", "feature_code": "VAL", "timzeone": "Asia/Seoul", "dem": "227", "admin1_code": "06", "asciiname": "Kaeun-dong", "alternativenames": "Kaeun-dong,gaeundong,\uac1c\uc6b4\ub3d9", "admin4_code": "", "population": "0"}},{"_index":"geonames","_type":"geoname","_id":"AU4wG06OSPt6k9TaGZ18","_score":1.0,"_source":{"elevation": "", "name": "Pomun-sa", "modification_date": "2014-01-01", "geonameid": "7584383", "feature_class": "S", "admin3_code": "", "admin2_code": "", "coordinates": "38.16296,128.21938", "cc2": "", "country_code3": "KOR", "country_code2": "KR", "feature_code": "TMPL", "timzeone": "Asia/Seoul", "dem": "356", "admin1_code": "06", "asciiname": "Pomun-sa", "alternativenames": "Pomun-sa,bomunsa,\ubcf4\ubb38\uc0ac", "admin4_code": "", "population": "0"}},{"_index":"geonames","_type":"geoname","_id":"AU4wG06OSPt6k9TaGZ2D","_score":1.0,"_source":{"elevation": "", "name": "Mokchang-gol", "modification_date": "2014-01-01", "geonameid": "7584524", "feature_class": "T", "admin3_code": "", "admin2_code": "", "coordinates": "38.2449,128.18576", "cc2": "", "country_code3": "KOR", "country_code2": "KR", "feature_code": "VAL", "timzeone": "Asia/Seoul", "dem": "477", "admin1_code": "06", "asciiname": "Mokchang-gol", "alternativenames": "Mokchang-gol,mogjang-gol,\ubaa9\uc7a5\uace8", "admin4_code": "", "population": "0"}},{"_index":"geonames","_type":"geoname","_id":"AU4wG06OSPt6k9TaGZ2I","_score":1.0,"_source":{"elevation": "", "name": "P\u2019iyang-gol", "modification_date": "2014-01-01", "geonameid": "7584529", "feature_class": "T", "admin3_code": "", "admin2_code": "", "coordinates": "38.21451,128.23218", "cc2": "", "country_code3": "KOR", "country_code2": "KR", "feature_code": "VAL", "timzeone": "Asia/Seoul", "dem": "503", "admin1_code": "06", "asciiname": "P'iyang-gol", "alternativenames": "P'iyang-gol,P\u2019iyang-gol,piyang-gol,\ud53c\uc591\uace8", "admin4_code": "", "population": "0"}},{"_index":"geonames","_type":"geoname","_id":"AU4wG06OSPt6k9TaGZ2N","_score":1.0,"_source":{"elevation": "", "name": "Sanghwang-gol", "modification_date": "2014-01-01", "geonameid": "7584534", "feature_class": "T", "admin3_code": "", "admin2_code": "", "coordinates": "38.19486,128.30389", "cc2": "", "country_code3": "KOR", "country_code2": "KR", "feature_code": "VAL", "timzeone": "Asia/Seoul", "dem": "377", "admin1_code": "06", "asciiname": "Sanghwang-gol", "alternativenames": "Sanghwang-gol,sanghwang-gol,\uc0c1\ud669\uace8", "admin4_code": "", "population": "0"}}]}}
root@linux-dev:~# 
root@linux-dev:~# uname -a
Linux linux-dev 4.4.0-31-generic #50-Ubuntu SMP Wed Jul 13 00:07:12 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
root@linux-dev:~# 

```

**Comment by ahalterman** [comment]
Thanks for catching that, @chrisbennight. I just pushed a commit to `adm1` that I think will resolve it. There was an issue with how the country code filtering was being done in the elastic query. I've switched it from a "filter" to a "query match". The downside of this approach is that it can only accept a single country filter rather than a list, but since the country picker only returns a single country, I don't think this is a big deal. When I re-do the elastic Dockerfile I can think about upgrading to elasticsearch 2.x, which should make it possible to [use a filter](http://elasticsearch-dsl.readthedocs.io/en/latest/search_dsl.html#filters) for multiple values. Let me know if that resolves it (and @Maryam-Imani, same).
